### PR TITLE
[SPARK-24966][SQL] Implement precedence rules for set operations.

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

Currently the set operations INTERSECT, UNION and EXCEPT are assigned the same precedence. This PR fixes the problem by giving INTERSECT  higher precedence than UNION and EXCEPT. UNION and EXCEPT operators are evaluated in the order in which they appear in the query from left to right.

This results in change in behavior because of the change in order of evaluations of set operators in a query. The old behavior is still preserved under a newly added config parameter.

Query `:`
```
SELECT * FROM t1
UNION 
SELECT * FROM t2
EXCEPT
SELECT * FROM t3
INTERSECT
SELECT * FROM t4
```
Parsed plan before the change `:`
```
== Parsed Logical Plan ==
'Intersect false
:- 'Except false
:  :- 'Distinct
:  :  +- 'Union
:  :     :- 'Project [*]
:  :     :  +- 'UnresolvedRelation `t1`
:  :     +- 'Project [*]
:  :        +- 'UnresolvedRelation `t2`
:  +- 'Project [*]
:     +- 'UnresolvedRelation `t3`
+- 'Project [*]
   +- 'UnresolvedRelation `t4`
```
Parsed plan after the change `:`
```
== Parsed Logical Plan ==
'Except false
:- 'Distinct
:  +- 'Union
:     :- 'Project [*]
:     :  +- 'UnresolvedRelation `t1`
:     +- 'Project [*]
:        +- 'UnresolvedRelation `t2`
+- 'Intersect false
   :- 'Project [*]
   :  +- 'UnresolvedRelation `t3`
   +- 'Project [*]
      +- 'UnresolvedRelation `t4`
```
## How was this patch tested?
Added tests in PlanParserSuite, SQLQueryTestSuite.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by holdensmagicalunicorn** [comment]
@dilipbiswal, thanks! I am a bot who has found some folks who might be able to help with the review:@gatorsmile, @rxin and @hvanhovell

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -17,6 +17,12 @@
 grammar SqlBase;
 
 @members {
+  /**
+   * When true, INTERSECT is given precedence over UNION and EXCEPT set operations as per
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: > When true, INTERSECT is given precedence over UNION and EXCEPT set operations as per

->

> When true, INTERSECT is given the greater precedence over the other set operations (UNION, EXCEPT and MINUS) as per

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 2]
```diff
@@ -676,4 +677,42 @@ class PlanParserSuite extends AnalysisTest {
...
+        |SELECT * FROM a
+        |UNION
+        |SELECT * FROM b
+        |EXCEPT ALL
+        |SELECT * FROM c
+        |INTERSECT ALL
+        |SELECT * FROM d
+      """.stripMargin
+
+    assertEqual(query1, Distinct(a.union(b)).except(c.intersect(d)))
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: also add `withSQLConf(SQLConf.SETOPS_PRECEDENCE_ENFORCED.key -> "true") {`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -1451,6 +1451,15 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val SETOPS_PRECEDENCE_ENFORCED =
+    buildConf("spark.sql.setops.precedence.enforced")
+      .doc("When set to true and order of evaluation is not specified by parentheses, " +
+        "INTERSECT operations are performed before any UNION or EXCEPT operations. " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: also include MINUS

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1451,6 +1451,15 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val SETOPS_PRECEDENCE_ENFORCED =
+    buildConf("spark.sql.setops.precedence.enforced")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: let me think about the name of conf

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -165,9 +165,9 @@ object SetOperation {
 }
 
 case class Intersect(
-   left: LogicalPlan,
-   right: LogicalPlan,
-   isAll: Boolean = false) extends SetOperation(left, right) {
+    left: LogicalPlan,
```
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: This is not related to the current PR. This addresses a comment from @HyukjinKwon in [21886](https://github.com/apache/spark/pull/21886)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 6]
```diff
@@ -535,14 +535,14 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case logical.Intersect(left, right, true) =>
         throw new IllegalStateException(
           "logical intersect operator should have been replaced by union, aggregate" +
-            "and generate operators in the optimizer")
+            " and generate operators in the optimizer")
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: This is not related to the current PR. This addresses a comment from @HyukjinKwon in [21886](https://github.com/apache/spark/pull/21886)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 7]
```diff
@@ -535,14 +535,14 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
       case logical.Except(left, right, false) =>
         throw new IllegalStateException(
           "logical except operator should have been replaced by anti-join in the optimizer")
       case logical.Except(left, right, true) =>
         throw new IllegalStateException(
           "logical except (all) operator should have been replaced by union, aggregate" +
-            "and generate operators in the optimizer")
+            " and generate operators in the optimizer")
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: This is not related to the current PR. This addresses a comment from @HyukjinKwon in [21886](https://github.com/apache/spark/pull/21886)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: @gatorsmile Sure.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: @gatorsmile A couple of alternatives i had considered :-). FYI.
`spark.sql.set.operators.precedence.enforced`
`spark.sql.set.operators.standard.compliance`

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@maropu Is this a transient failure ? Does not seem related to my change ?

**Comment by maropu** [comment]
no idea, but `HiveClientSuites` seems flaky: https://issues.apache.org/jira/browse/SPARK-23622 (the error message is different though...)

**Comment by maropu** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@gatorsmile rebasing in a hope that the test result changes :-)

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM 

Thanks! Merged to master.

**Comment by dilipbiswal** [comment]
Thanks a lot @gatorsmile
